src/core/reference/src/op/paged_attention.cpp

```diff
@@ -0,0 +1,336 @@
+    const T* value,                       // shape: [batch_tokens, num_kv_heads * head_size]
+    const T* key_cache,                   // shape: [num_blocks, num_kv_heads, block_size, head_size]
+    const T* value_cache,                 // shape: [num_blocks, num_kv_heads, block_size, head_size]
+    const Shape& q_shape,                 // e.g. {batch_tokens, num_heads * head_size}
+    const Shape& kv_shape,                // e.g. {batch_tokens, num_kv_heads * head_size}
+    const Shape& kv_cache_shape,          // e.g. {num_blocks, num_kv_heads, block_size, head_size}
+    const int32_t* past_lens,             // [batch_seq]: past tokens per sequence
+    const Shape& past_lens_shape,         // e.g. {batch_seq}
+    const int32_t* subsequence_begins,    // [batch_seq + 1]: start indices of new tokens per sequence
+    const int32_t* block_indices,         // [num_blocks]: block table for each sequence
+    const int32_t* block_indices_begins,  // [batch_seq + 1]: indices into block_indices per sequence
+    const T* scale_ptr,                   // (Optional) attention scale factor (can be nullptr; default = 1)
+    const int32_t* sliding_window_ptr,    // (Optional) sliding window parameter (can be nullptr; default = 0)
+    const T* alibi_slopes,  // (Optional) [num_kv_heads]: per-head bias slopes (can be nullptr; default = 0)
+    const int32_t* max_context_len_ptr,  // max context length (for score output indexing)
+
+    // Rotation parameters (if any is nullptr, rotation is skipped)
+    const int32_t* rotated_block_indices,  // (Optional) [num_rotated_blocks]: blocks to which RoPE is applied
+    const int32_t* rotation_deltas,  // (Optional) [num_rotated_blocks, block_size || 1]: indices into the trig LUT
+    const T* rotation_trig_lut,      // (Optional) LUT: [lut_rows, head_size] (first half: cosines, second half: sines)
+    const Shape& rotated_block_indices_shape,  // shape of rotated_block_indices (e.g. {num_rotated_blocks})
+    const Shape& rotation_deltas_shape,        // shape of rotation_deltas (e.g. {num_rotated_blocks, block_size} or
+                                               // {num_rotated_blocks, 1})
+    const Shape& rotation_trig_lut_shape) {    // shape of rotation_trig_lut
+
+    // Use default values if pointers are null.
+    T scale = scale_ptr ? scale_ptr[0] : T(1);
+    int sliding_window = sliding_window_ptr ? sliding_window_ptr[0] : 0;
+    int max_context_len = max_context_len_ptr ? max_context_len_ptr[0] : 0;
+
+    // Determine dimensions.
+    int num_blocks = kv_cache_shape[0];
+    int num_kv_heads = kv_cache_shape[1];
+    int block_size = kv_cache_shape[2];
+    int head_size = kv_cache_shape[3];
+
+    int batch_tokens = q_shape[0];
+    int query_features = q_shape[1];  // equals num_heads * head_size.
+    int num_heads = query_features / head_size;
+    int batch_seq = past_lens_shape.empty() ? 1 : past_lens_shape[0];
```

Comment:
Usage of `int` should be ok for most of the platforms but to ensure stability, I would recommend to specify the size of the integer explicitly (int32_t) or (int64_t). I can see the proposed reference mostly rely on i32 type, it's ok if higher precision is not used/requested. In reference implementations int64_t is often use to handle both i32 and i64 at once (and reduce risk of overflow).